File: HOT/Array/LeetCode34.go
package array

func searchRange(nums []int, target int) []int {
	if nums == nil {
		return []int{-1, -1}
	}
	lower := lower_search(nums, target)
	upper := upper_search(nums, target) - 1

	if lower == len(nums) || nums[lower] != target {
		return []int{-1, -1}
	}
	return []int{lower, upper}

}

func lower_search(nums []int, target int) int {
	l, r := 0, len(nums)
	for l < r {
		mid := l + (r-l)/2
		if nums[mid] >= target {
			r = mid
		} else {
			l = mid + 1
		}
	}
	return l
}

func upper_search(nums []int, target int) int {
	l, r := 0, len(nums)
	for l < r {
		mid := l + (r-l)/2
		if nums[mid] > target {
			r = mid
		} else {
			l = mid + 1
		}
	}
	return l
}

File: HOT/Array/LeetCode34.cc
#include<iostream>
#include<vector>
using namespace std;

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        if(nums.empty()){
            return {-1,-1};
        }
        int lower = lower_serarch(nums,target);
        int upper = upper_search(nums,target)-1;
        if (lower == nums.size() || nums[lower] != target){
            return {-1,-1};
        }
        return {lower,upper};
    }
    //[L,R)
    int lower_serarch(vector<int>& nums, int target){
      int l = 0,r= nums.size();
      while(l < r){
          int mid = l+(r-l) /2;
          if(nums[mid] >= target){
              r = mid;
          }
          else{
              l = mid+1;
          }
      }
      return l;  
    }

    //[L,R]

    int upper_search(vector<int>& nums, int target){
    int l = 0,r = nums.size();
      while(l < r){
          int mid = l+(r-l) /2;
          if(nums[mid] > target){
              r = mid;
          }
          else{
              l = mid+1;
          }
      }
      return l;
    }
};